cumulus/parachains/runtimes/testing/penpal/src/pallets/transact_helper.rs

```diff
@@ -0,0 +1,147 @@
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_core::{H160, H256};
+	use sp_std::{boxed::Box, vec, vec::Vec};
+	use xcm::prelude::*;
+	use xcm_executor::traits::XcmAssetTransfers;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_xcm::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type XcmRouter: SendXcm;
+		type XcmExecutor: ExecuteXcm<<Self as frame_system::Config>::RuntimeCall>
+			+ XcmAssetTransfers;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// ExportMessage was sent
+		SentExportMessage { message_id: XcmHash, sender_cost: Assets, message: Xcm<()> },
+		/// XCM message sent. \[to, message\]
+		Sent { to: Location, message: Xcm<()> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidMsg,
+		FeesNotMet,
+		SendFailure,
+		BadVersion,
+		Unreachable,
+	}
+
+	#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+	pub struct TransactInfo {
+		pub agent_id: H256,
+		pub target: H160,
+		pub call: Vec<u8>,
+		pub gas_limit: u64,
+		pub fee: u128,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<<T as frame_system::Config>::AccountId>,
+	{
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::from_parts(100_000_000, 0))]
+		pub fn send_as_sovereign(
+			origin: OriginFor<T>,
+			dest: Box<VersionedLocation>,
+			message: Box<VersionedXcm<()>>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let dest = Location::try_from(*dest).map_err(|()| Error::<T>::BadVersion)?;
+			let message: Xcm<()> = (*message).try_into().map_err(|()| Error::<T>::BadVersion)?;
+
+			pallet_xcm::Pallet::<T>::send_xcm(Here, dest.clone(), message.clone()).map_err(
+				|e| match e {
+					SendError::Unroutable => Error::<T>::Unreachable,
+					_ => Error::<T>::SendFailure,
+				},
+			)?;
+			Self::deposit_event(Event::Sent { to: dest, message });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::from_parts(100_000_000, 0))]
+		pub fn transact_to_ethereum(
+			origin: OriginFor<T>,
+			agent_id: H256,
+			target: H160,
+			call: Vec<u8>,
+			fee: u128,
+			gas_limit: u64,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let dest = Location {
+				parents: 2,
+				interior: Junctions::from([GlobalConsensus(Ethereum { chain_id: 11155111 })]),
+			};
+
+			let transact = TransactInfo { agent_id, target, call, gas_limit, fee };
+
+			let inner_message = Xcm(vec![
+				Transact {
+					origin_kind: OriginKind::SovereignAccount,
+					require_weight_at_most: Weight::default(),
+					call: transact.encode().into(),
+				},
+				SetTopic([0; 32]),
+			]);
+
+			let (ticket, price) =
```

Comment:
Using XCMRouter directly and manually collecting fees is too low level.

We can use `XcmExecutor::execute` with an `ExportMessage` instruction:
```
ExportMessage {
  dest = GlobalConsensus(Ethereum { .. })
  xcm = {
    Transact {origin_kind: OriginKind::SovereignAccount,},
    SetTopic([0; 32]),
  }
}
```

Or we can use `PalletXcm::send_xcm` which will automatically collect fees from the User.

Reply:
For the transact there are 2 portions of the fees collected from the User(local on BH&remote on Ethereum). For the fee on Ethereum is dynamic as the parameter of the extrinsic.  

So we need explicit `charge_fees` including the both.

https://github.com/Snowfork/polkadot-sdk/blob/6cc2175ff5063178260cfa5e243aa981d32a3622/cumulus/parachains/runtimes/testing/penpal/src/pallets/transact_helper.rs#L128-L129

Anyway, this is just a refence impl for the transact on 3rd chain. They can choose whatever fee policy as they want.